### PR TITLE
[feature](function) Support for aggregate function foreach combiner for some error function

### PR DESCRIPTION
**Pull request by Mryange**
## Proposed changes

```
mysql [test]>select * from foreach_table;
+------+-----------+-----------------------+-------------------------+
| id   | a         | b                     | s                       |
+------+-----------+-----------------------+-------------------------+
|    1 | [1, 2, 3] | [[1], [1, 2, 3], [2]] | ["ab", "123", "114514"] |
|    2 | [20]      | [[2]]                 | ["cd"]                  |
|    3 | [100]     | [[1]]                 | ["efg"]                 |
|    4 | NULL      | [null]                | NULL                    |
|    5 | [null, 2] | [[2], null]           | [null, "c"]             |
+------+-----------+-----------------------+-------------------------+
5 rows in set (0.01 sec)

mysql [test]>select array_agg_foreach(s) from foreach_table;
+-------------------------------------------------------+
| array_agg_foreach(s)                                  |
+-------------------------------------------------------+
| [["ab", "cd", "efg", null], ["123", "c"], ["114514"]] |
+-------------------------------------------------------+
1 row in set (0.02 sec)

mysql [test]>select map_agg_foreach(s,a) from foreach_table;
+----------------------------------------------------------------+
| map_agg_foreach(s, a)                                          |
+----------------------------------------------------------------+
| [{"ab":1, "cd":20, "efg":100}, {"123":2, "c":2}, {"114514":3}] |
+----------------------------------------------------------------+
1 row in set (0.02 sec)
```

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

**Review by HappenLee — DISMISSED** [review]
LGTM

File: be/src/vec/aggregate_functions/aggregate_function_collect.h [hunk 1]
```diff
@@ -491,25 +574,23 @@ class AggregateFunctionCollect
     void merge(AggregateDataPtr __restrict place, ConstAggregateDataPtr rhs,
                Arena* arena) const override {
         auto& data = this->data(place);
-        auto& rhs_data = this->data(rhs);
+        const auto& rhs_data = this->data(rhs);
         if constexpr (ENABLE_ARENA) {
             data.merge(rhs_data, arena);
         } else if constexpr (!ShowNull::value) {
             data.merge(rhs_data);
```
Comment (BiteTheDDDDt) [inline be/src/vec/aggregate_functions/aggregate_function_collect.h:581]: why two branch have same code

File: be/src/vec/aggregate_functions/aggregate_function_map.h [hunk 2]
```diff
@@ -135,13 +135,44 @@ struct AggregateFunctionMapAggData {
...
+    }
+
+    void read(BufferReadable& buf) {
+        size_t size = 0;
+        read_binary(size, buf);
+        std::string keys;
+        for (size_t i = 0; i < size; i++) {
+            read_binary(keys, buf);
+            StringRef key {keys};
+            if (_map.find(key) != _map.cend()) {
```
Comment (BiteTheDDDDt) [inline be/src/vec/aggregate_functions/aggregate_function_map.h:163]: when will the key not be found?

Comment (Mryange) [reply be/src/vec/aggregate_functions/aggregate_function_map.h:163]: do nothing， the map aggregation function assumed that keys cannot be duplicated, so if there are duplicates, they should not be placed in the map.

**Comment by Mryange** [comment]
run buildall

**Review by BiteTheDDDDt — DISMISSED** [review]

**Comment by Mryange** [comment]
run buildall

File: be/src/vec/aggregate_functions/aggregate_function_collect.h [hunk 3]
```diff
@@ -362,12 +364,58 @@ struct AggregateFunctionArrayAggData {
...
+            write_binary(nested_column->get_data()[i], buf);
+        }
+    }
+
+    void read(BufferReadable& buf) {
+        DCHECK(null_map);
+        DCHECK(null_map->empty());
+        size_t size = 0;
+        read_binary(size, buf);
+        null_map->reserve(size);
```
Comment (HappenLee) [inline be/src/vec/aggregate_functions/aggregate_function_collect.h:386]: better use resize? 
and null_map[i] = null_value;

File: be/src/vec/aggregate_functions/aggregate_function_collect.h [hunk 4]
```diff
@@ -417,6 +465,49 @@ struct AggregateFunctionArrayAggData<StringRef> {
...
+        nested_column->reserve(size);
+
+        NullMap::value_type null_value;
+        for (size_t i = 0; i < size; i++) {
+            read_binary(null_value, buf);
+            null_map->push_back(null_value);
+        }
+
+        std::string s;
```
Comment (HappenLee) [inline be/src/vec/aggregate_functions/aggregate_function_collect.h:493]: better use `StringRef ` to reduce the memcpy

File: be/src/vec/aggregate_functions/aggregate_function_map.h [hunk 5]
```diff
@@ -135,13 +135,44 @@ struct AggregateFunctionMapAggData {
...
+                continue;
+            }
+            key.data = _arena.insert(key.data, key.size);
+            assert_cast<KeyColumnType&>(*_key_column).insert_data(key.data, key.size);
+        }
+        std::string val;
```
Comment (HappenLee) [inline be/src/vec/aggregate_functions/aggregate_function_map.h:169]: same to up

**Comment by Mryange** [comment]
run buildall

**Review by HappenLee — APPROVED** [review]
LGTM

**Review by HappenLee — APPROVED** [review]
LGTM

**Comment by Mryange** [comment]
run buildall

**Review by BiteTheDDDDt — APPROVED** [review]

**Comment by Mryange** [comment]
run p0

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

**Pull request merged**